File: backend/src/main/java/be/technobel/parsemaster/service/implementation/AuthServiceImpl.java
package be.technobel.parsemaster.service.implementation;

import be.technobel.parsemaster.dto.AuthDTO;
import be.technobel.parsemaster.exception.Exceptions;
import be.technobel.parsemaster.form.ChangePasswordForm;
import be.technobel.parsemaster.form.LoginForm;
import be.technobel.parsemaster.form.RequestPasswordForm;
import be.technobel.parsemaster.repository.UserRepository;
import be.technobel.parsemaster.security.JWTProvider;
import be.technobel.parsemaster.service.declaration.AuthService;
import be.technobel.parsemaster.service.declaration.EmailService;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import java.time.Instant;
import java.util.Comparator;
import java.util.TreeSet;

@Service
public class AuthServiceImpl implements AuthService {
  private final UserRepository userRepository;
  private final AuthenticationManager authenticationManager;
  private final JWTProvider jwtProvider;
  private final PasswordEncoder passwordEncoder;
  private final UidService uidService;
  private final EmailService emailService;

  private record PasswordRequest(String token, String email, Instant expiration) {}
  private final long PASSWORD_REQUEST_EXPIRATION = 1000L*60*5; // milliseconds
  private final String PASSWORD_REQUEST_URL;
  private final TreeSet<PasswordRequest> passwordRequests = new TreeSet<>(
    Comparator.comparing(PasswordRequest::expiration)
  );

  public AuthServiceImpl(
    @Value("${spring.profiles.active}") String activeProfile,
    @Value("${app.http-host}") String httpHost,
    UserRepository userRepository,
    AuthenticationManager authenticationManager,
    JWTProvider jwtProvider,
    PasswordEncoder passwordEncoder,
    UidService uidService,
    EmailService emailService
  ) {
    this.userRepository = userRepository;
    this.authenticationManager = authenticationManager;
    this.jwtProvider = jwtProvider;
    this.passwordEncoder = passwordEncoder;
    this.uidService = uidService;
    this.emailService = emailService;
    if(activeProfile.equals("prod")) {
      this.PASSWORD_REQUEST_URL = httpHost+"/auth/changePassword/";
    } else {
      this.PASSWORD_REQUEST_URL = "http://localhost:4200/auth/changePassword/";
    }
  }

  @Override
  public AuthDTO signIn(LoginForm form) {
    try {
      authenticationManager.authenticate(
        new UsernamePasswordAuthenticationToken(form.login(), form.password())
      );
      final var user = userRepository
        .findByEmailOrPseudonym(form.login())
        .orElseThrow(Exceptions.USER_NOT_FOUND::create);
      return new AuthDTO(
        jwtProvider.generateToken(user.getUsername(), user.getRole())
      );
    }
    catch(Exception ex) {
      throw Exceptions.BAD_CREDENTIALS.create();
    }
  }

  @Override
  public void signOut(HttpServletRequest request) {
    final var token = jwtProvider.extractToken(request);
    if(token != null && jwtProvider.validateToken(token)) jwtProvider.revoke(token);
    else throw Exceptions.BAD_TOKEN.create();
  }

  @Override
  public void requestPassword(RequestPasswordForm form) {
    cleanPasswordTokens();
    final var email = form.email();
    final var notExists = passwordRequests.stream().noneMatch(t -> t.email().equals(email));
    if(notExists) {
      final var token = uidService.generateId("request-password");
      passwordRequests.add(new PasswordRequest(
        token, email, Instant.now().plusMillis(PASSWORD_REQUEST_EXPIRATION)
      ));
      emailService.passwordModificationRequest(email, PASSWORD_REQUEST_URL+token);
    }
  }

  @Override
  public void changePassword(ChangePasswordForm form, String token) {
    cleanPasswordTokens();
    final var request = passwordRequests.stream().filter(t -> t.token.equals(token)).findFirst()
      .orElseThrow(Exceptions.TOKEN_NOT_FOUND::create);
    final var user = userRepository.findByEmail(request.email())
      .orElseThrow(Exceptions.USER_NOT_FOUND::create);
    user.setPasswordHash(passwordEncoder.encode(form.password()));
    userRepository.save(user);
    passwordRequests.remove(request);
    emailService.passwordModificationConfirmation(request.email());
  }

  @Override
  public boolean existsPasswordToken(String token) {
    cleanPasswordTokens();
    return passwordRequests.stream().anyMatch(t -> t.token.equals(token));
  }

  private void cleanPasswordTokens() {
    final var now = Instant.now();
    passwordRequests.removeIf(revoked -> revoked.expiration().isBefore(now));
  }
}

File: src/main/kotlin/be/technobel/playzone/bll/services/impl/AuthServiceImpl.kt
package be.technobel.playzone.bll.services.impl

import be.technobel.playzone.bll.exceptions.NotFoundException
import be.technobel.playzone.bll.services.AuthService
import be.technobel.playzone.bll.services.EmailService
import be.technobel.playzone.dal.repositories.UserRepository
import be.technobel.playzone.pl.models.forms.LoginForm
import be.technobel.playzone.pl.security.JWTProvider
import jakarta.servlet.http.HttpServletRequest
import org.springframework.core.env.Environment
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.BadCredentialsException
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.AuthenticationException
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import java.time.Instant
import java.util.*

/** Durée de vie en millisecondes d'un token */
private const val PASSWORD_TOKEN_EXPIRATION = 1000L*60*5

@Service
class AuthServiceImpl(
	private val userRepository: UserRepository,
	private val authenticationManager: AuthenticationManager,
	private val jwtProvider: JWTProvider,
	private val passwordEncoder: PasswordEncoder,
	private val uidService: UidService,
	private val emailService: EmailService,
	private val environment: Environment,
) : AuthService {
	private data class PasswordToken(val token: String, val email: String, val expiration: Instant)
	private val passwordTokens = TreeSet<PasswordToken> { a, b -> a.expiration.compareTo(b.expiration) }

	@Throws(AuthenticationException::class)
	override fun login(form: LoginForm): String {
		try {
			authenticationManager.authenticate(UsernamePasswordAuthenticationToken(form.login, form.password))
			val user = userRepository.findByEmail(form.login!!)!!
			return jwtProvider.generateToken(user.username, user.role)
		}
		catch (ex: AuthenticationException) {
			throw ex
		}
	}

	@Throws(BadCredentialsException::class)
	override fun logout(request: HttpServletRequest) {
		val token = jwtProvider.extractToken(request)
		?: throw BadCredentialsException("Invalid token")
		if(jwtProvider.validateToken(token)) jwtProvider.revoke(token)
		else throw BadCredentialsException("Invalid token")
	}

	override fun passwordModificationRequest(email: String) {
		cleanPasswordTokens()
		if(!passwordTokens.any { it.email == email }) {
			val token = uidService.generateBase64()
			passwordTokens.add(PasswordToken(
				token, email, Instant.now().plusMillis(PASSWORD_TOKEN_EXPIRATION)
			))
			if(environment.activeProfiles.contains("production")) {
				emailService.sendPasswordModificationEmail(
					email, "http://172.20.50.19:4200/auth/changePassword/${token}"
				)
			} else {
				emailService.sendPasswordModificationEmail(
					email, "http://localhost:4200/auth/changePassword/${token}"
				)
			}

		}
	}

	@Throws(NotFoundException::class)
	override fun passwordModification(token: String, password: String) {
		cleanPasswordTokens()
		val passwdToken = passwordTokens.find { it.token == token }
		?: throw NotFoundException("Token not found")
		val user = userRepository.findByEmail(passwdToken.email)
		?: throw NotFoundException("User not found")
		userRepository.modifyPasswordByUid(user.uid, passwordEncoder.encode(password))
		emailService.sendPasswordModificationConfirmationEmail(passwdToken.email)
		passwordTokens.remove(passwdToken)
	}

	/** Supprime les tokens expirés */
	private fun cleanPasswordTokens() {
		val now = Instant.now()
		passwordTokens.removeIf { it.expiration < now }
	}
}